src/main/java/org/veupathdb/service/eda/ss/model/reducer/formatter/MultiValueFormatter.java

```diff
@@ -2,32 +2,34 @@
 
 import org.veupathdb.service.eda.ss.model.reducer.FormattedTabularRecordStreamer;
 
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
 import java.util.List;
 import java.util.stream.Collectors;
 
 public class MultiValueFormatter implements TabularValueFormatter {
 
   @Override
-  public String format(List<String> values) {
-    StringBuilder valueRecord = new StringBuilder("[");
-    valueRecord.append(values.stream().collect(Collectors.joining(",")));
-    valueRecord.append("]");
-    return valueRecord.toString();
-  }
-
-  @Override
-  public String format(FormattedTabularRecordStreamer.ValueStream<String> stream,
+  public byte[] format(FormattedTabularRecordStreamer.ValueStream<byte[]> stream,
                        long idIndex) {
-    StringBuilder valueRecord = new StringBuilder("[");
-    boolean first = true;
+    List<byte[]> allRecords = new ArrayList<>();
     while (stream.hasNext() && stream.peek().getIdIndex() == idIndex) {
-      if (!first) {
-        valueRecord.append(",");
+      allRecords.add(stream.next().getValue());
+    }
+    final int numCommas = allRecords.size() - 1;
+    final byte[] result = new byte[numCommas + allRecords.stream().mapToInt(arr -> arr.length).sum() + 2];
+    result[0] = '[';
+    result[result.length - 1] = ']';
+    int index = 1;
+    int recordCount = 0;
+    for (byte[] record: allRecords) {
```

Comment:
Convert to regular for loop since you're doing all the things?  Saves duplicated iteration logic and ArrayList.get() is should be O(1).

Reply:
Unfortunately, the index out of bounds check in `ArrayList.get` has given me trouble in these inner loops before. I'll give it a try to see if it makes a difference in practice in this case.